### PR TITLE
[espnet3-3] Add trainer and model

### PR DESCRIPTION
**Pull request by Masao-Someki**
## What did you change?

Added the main training script for ESPnet-3, including the following components:
- trainer.py
- model.py
- dataloader.py (split from model.py due to its large functionality)

---

## Why did you make this change?

To support training in ESPnet-3 using PyTorch Lightning. This implementation is based on Yifan’s ESPnet-2 Lightning trainer and has been updated to offer a more flexible interface.

---

## Is your PR small enough?

Yes, the core content of this PR is focused and concise. Although it includes a significant number of additions, the code is well-documented and thoroughly tested.
Note that this PR depends on the following related PRs:

#6139
#6167
#6171

Once these are merged, the total number of affected files will be 13.

---

## Additional Context

#6133

### CODE REVIEW

**Comment by Masao-Someki** [comment]
## Test Case for ESPnet3LightningTrainer


**Hydra-based Component Initialization**
| Test Name                  | Description                                                                  |
|---------------------------|------------------------------------------------------------------------------|
| test_logger_variants      | logger config (None or dict) correctly instantiates a TensorBoardLogger      |
| test_accelerator_variants | accelerator config (None or dict) correctly sets CPUAccelerator              |
| test_strategy_variants    | strategy config (None or DDP) sets correct training strategy                 |
| test_profiler_variants    | profiler config is properly instantiated (PassThrough or SimpleProfiler)     |
| test_plugins_variants     | plugins config correctly sets HalfPrecision plugin or None                   |

**Config Overrides Based on ESPnet3 Convention**
| Test Name                              | Description                                                                  |
|----------------------------------------|------------------------------------------------------------------------------|
| test_reload_dataloaders_enforced       | Forces `reload_dataloaders_every_n_epochs=1` if `iter_factory` is enabled   |
| test_distributed_sampler_disabled_if_espnet_sampler | Forces `use_distributed_sampler=False` when ESPnet sampler is used     |

**Method Delegation (fit / validate)**
| Test Name                    | Description                                                                  |
|-----------------------------|------------------------------------------------------------------------------|
| test_fit_calls_trainer_fit  | Ensures `fit()` forwards args/kwargs to `trainer.fit()` with model injection |
| test_validate_calls_trainer_validate | Ensures `validate()` forwards args/kwargs to `trainer.validate()`        |

**Assertion & Required Fields**
| Test Name                    | Description                                                                  |
|-----------------------------|------------------------------------------------------------------------------|
| test_missing_model_raises   | Raises `TypeCheckError` if `model` is None                                   |
| test_missing_expdir_raises  | Raises `TypeCheckError` if `expdir` is None                                  |
| test_missing_config_raises  | Raises `TypeCheckError` if `config` is None                                  |

**Comment by Masao-Someki** [comment]
## Test Case for LitESPnetModel


**Normal Cases**
| Test Name                          | Description                                                              |
|-----------------------------------|--------------------------------------------------------------------------|
| test_training_step_runs           | Verifies that training_step executes and returns a loss                  |
| test_validation_step_runs         | Verifies that validation_step runs and returns a loss                    |
| test_is_espnet_sampler_flag       | Checks that is_espnet_sampler is set correctly based on iter_factory     |
| test_state_dict_and_load          | Ensures state_dict and load_state_dict are correctly forwarded           |
| test_use_espnet_collator_flag_false | Checks dataset.use_espnet_collator=False for non-CommonCollateFn        |

**Error / Edge Cases**
| Test Name                          | Description                                                              |
|-----------------------------------|--------------------------------------------------------------------------|
| test_nan_loss_skip                | Ensures batch is skipped when loss is NaN/Inf                            |
| test_dataloader_mismatch_raises  | Asserts error if train/valid dataloader types mismatch                   |
| test_mixed_optim_scheduler_raises| Asserts error if both optim and optims/schedulers are defined            |
| test_missing_optimizer_and_scheduler_raises | Raises ValueError when neither optimizer nor scheduler is specified     |

**Comment by Masao-Someki** [comment]
## Test Case for DataLoaderBuilder


**Standard PyTorch DataLoader mode**
| Test Name                          | Description                                                              |
|-----------------------------------|--------------------------------------------------------------------------|
| test_batch_sampler_only           | Uses only batch_sampler and omits batch_size and shuffle                |
| test_sampler_only                 | Uses only sampler to drive the dataloader                               |
| test_collate_fn_none              | Sets collate_fn=None to use PyTorch's default_collate                   |
| test_common_collate_fn            | Uses CommonCollateFn and checks audio/audio_lengths formatting          |
| test_custom_collate_fn            | Uses a custom collate function and checks the batch structure           |
| test_sampler_and_batch_sampler_conflict | Ensures AssertionError if both sampler and batch_sampler are defined  |

**IterFactory Mode & YAML-based Integration**
| Test Name                          | Description                                                              |
|-----------------------------------|--------------------------------------------------------------------------|
| test_iter_factory_from_default_yaml_with_organizer | Uses a real YAML config with iter_factory and verifies batch structure |
| test_iter_factory_with_collate_fn | Confirms config-defined collate_fn takes precedence over argument       |

**Multiple Iterator Mode (Sharded Dataset)**
| Test Name                              | Description                                                              |
|---------------------------------------|--------------------------------------------------------------------------|
| test_multiple_iterator_shard_initialization | Checks correct shard is selected based on epoch                         |
| test_multiple_iterator_epoch_shard_switching | Ensures different shard is selected as epoch changes                  |

File: espnet3/trainer/model.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
+import logging
+
+import lightning as L
```
Comment (sw005320) [inline espnet3/trainer/model.py]: Do you need to make it as `L`?
It is not frequently used, and `lightning` is more explicit

File: espnet3/trainer/trainer.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
+import warnings
+from argparse import Namespace
+from typing import Any, Dict, Union
+
+import lightning as L
```
Comment (sw005320) [inline espnet3/trainer/trainer.py]: ditto
Do you need to make it as `L`?
It is not frequently used, and `lightning` is more explicit

File: espnet3/trainer/model.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+from espnet3.trainer.dataloader import DataLoaderBuilder
+
+logger = logging.getLogger("lightning")
+
+
+class LitESPnetModel(L.LightningModule):
+    """
+    ESPnet3-specific PyTorch LightningModule wrapper.
+
+    This class handles model training, validation, optimizer/scheduler setup,
```
Comment (sw005320) [inline espnet3/trainer/model.py:23]: add a description about organizer

File: espnet3/trainer/model.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        valid_dataset: Validation dataset.
+        collate_fn (Callable): Collation function used in DataLoader.
+        is_espnet_sampler (bool): Whether the model uses ESPnet's custom sampler.
+    """
+
+    def __init__(self, model, config):
+        super().__init__()
+        self.config = config
+        self.model = model
+        organizer = instantiate(config.dataset)
```
Comment (sw005320) [inline espnet3/trainer/model.py]: I think `organizer` is a bit confusing.
Can you rename it?

File: espnet3/trainer/model.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
+import logging
+
+import lightning
+import torch
+from hydra.utils import instantiate
+from omegaconf import OmegaConf
+
+from espnet2.train.collate_fn import CommonCollateFn
+
+# Temporarily disabled for the code review.
```
Comment (popcornell) [inline espnet3/trainer/model.py:10]: what does this mean ? Is it because you want this PR to be independent ?

File: espnet3/trainer/model.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        ), "Train and valid should have the same type of dataloader."
+
+        self.is_espnet_sampler = is_train_espnet
+
+        self.collate_fn = CommonCollateFn(int_pad_value=-1)
+        # define collate_fn. Default to ESPnet's CommonCollateFn
+        if hasattr(self.config.dataloader, "collate_fn"):
+            self.collate_fn = instantiate(self.config.dataloader.collate_fn)
+
+    def _sync2skip(self, flag_skip):
```
Comment (popcornell) [inline espnet3/trainer/model.py:78]: Another way to do this is https://github.com/Lightning-AI/pytorch-lightning/issues/4956#issuecomment-1722500885 which will check also for invalid gradients.

Comment (popcornell) [reply espnet3/trainer/model.py:78]: However it will need some testing for ddp. (I guess it should work because it calls self.zero_grad())

File: espnet3/trainer/model.py (hunk 5 above)
Comment (sw005320) [reply espnet3/trainer/model.py:10]: Yes.
He split several PRs and needs this treatment.
Please ignore them.

File: espnet3/trainer/model.py (hunk 6 above)
Comment (Masao-Someki) [reply espnet3/trainer/model.py:78]: Thank you! This was copied from Yifan's Lightning model module in ESPnet2. Since it's already in use there, I didn't want to introduce too many changes. The implementation looks good, so I'll leave this item in the load map for now—we can revisit and implement it later.

Comment (sw005320) [reply espnet3/trainer/model.py:78]: Thanks, @popcornell.
Yeah, ddp test is a bit complicated, and we may take this option, but we may revisit once espnet3 recipe becomes more stable.

File: espnet3/trainer/model.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        self.valid_dataset = data_organizer.valid
+        self.nan_countdown = 0
+
+        # If user is trying to use both Pytorch dataloader and ESPnet's dataloader
+        # Then raise an error here.
+        is_train_espnet = False
+        if (
+            hasattr(self.config.dataloader.train, "iter_factory")
+            and self.config.dataloader.train.iter_factory is not None
+        ):
```
Comment (Emrys365) [inline espnet3/trainer/model.py:57]: Can be shortened as

```python
        if getattr(self.config.dataloader.train, "iter_factory", None) is not None:
```

File: espnet3/trainer/model.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+            is_train_espnet = True
+
+        is_valid_espnet = False
+        if (
+            hasattr(self.config.dataloader.valid, "iter_factory")
+            and self.config.dataloader.valid.iter_factory is not None
+        ):
```
Comment (Emrys365) [inline espnet3/trainer/model.py:64]: Can be shortened as

```python
        if getattr(self.config.dataloader.valid, "iter_factory", None) is not None:
```

File: espnet3/trainer/model.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+
+        any_invalid = self._check_nan_inf_loss(loss, batch_idx)
+        if any_invalid:
+            # skip this batch altogether on all workers.
+            return None
+
+        new_stats = {}
+        for k, v in stats.items():
+            if v is not None:
+                new_stats[f"{mode}/{k}"] = v.item()
```
Comment (Emrys365) [inline espnet3/trainer/model.py:165]: Here, is `v` always a Tensor? If it belongs to any other type (e.g., float), this line will fail.

File: espnet3/trainer/trainer.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        # Since espnet's sampler requires to set the following configs:
+        # Reload dataloaders every epoch to reuse ESPnet's dataloader
+        # reload_dataloaders_every_n_epochs=1
+        # ESPnet's dataloader already shards the dataset based on distributed setups
+        # use_distributed_sampler=False
+        if hasattr(self.model.config.dataloader.train, "iter_factory"):
+            if (
+                hasattr(self.config, "reload_dataloaders_every_n_epochs")
+                and self.config.reload_dataloaders_every_n_epochs != 1
+            ):
```
Comment (Emrys365) [inline espnet3/trainer/trainer.py:114]: Can be shortened as

```python
        if getattr(self.config, "reload_dataloaders_every_n_epochs", 1) != 1:
```

File: espnet3/trainer/trainer.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+                warnings.warn(
+                    "ESPnet's dataloader requires to set"
+                    "reload_dataloaders_every_n_epochs = 1. "
+                    "Override the config to 1."
+                )
+            if (
+                hasattr(self.model.config, "use_distributed_sampler")
+                and not self.config.use_distributed_sampler
+            ):
```
Comment (Emrys365) [inline espnet3/trainer/trainer.py:123]: Why using both `self.model.config` and `self.config` here? Are they the same object?

**Comment by sw005320** [comment]
Thanks, @Emrys365.

@Masao-Someki, can you reflect them by making a follow-up email?
